paasta_tools/cleanup_kubernetes_jobs.py

```diff
@@ -95,44 +104,118 @@ def alert_state_change(application: Application, soa_dir: str) -> Generator:
         raise
 
 
+def instance_is_not_bouncing(
+    instance_config: KubernetesDeploymentConfig,
+    applications: List[Application],
+    kube_client: KubeClient,
+) -> Optional[bool]:
+    for application in applications:
+        if isinstance(application, DeploymentWrapper):
+            existing_app = application.get_existing_app(kube_client)
+            if (
+                application.kube_deployment.namespace == instance_config.get_namespace()
+                and (
+                    instance_config.config_dict.get("instances", None)
+                    == existing_app.status.ready_replicas
+                    or instance_config.config_dict.get("min_instances", None)
+                    <= existing_app.status.ready_replicas
+                )
+            ):
+                return True
+            else:
+                return False
+
+        elif isinstance(application, StatefulSetWrapper):
+            log.critical(
+                "Paasta detected a StatefulSet that was migrated to a new namespace"
+                "StatefulSet bouncing across namespaces is not supported"
+            )
+            raise StatefulSetsAreNotSupportedError
+    return None
+
+
+def get_applications_to_kill(
+    applications_dict: Dict[Tuple[str, str], List[Application]],
+    cluster: str,
+    valid_services: Set[Tuple[str, str]],
+    kube_client: KubeClient,
+    soa_dir: str,
+) -> List[Application]:
+    """
+
+    :param applications_dict: A dictionary with (service, instance) as keys and a list of applications for each tuple
+    :param cluster: paasta cluster
+    :param valid_services: a set with the valid (service, instance) tuples for this cluster
+    :param kube_client:
+    :param soa_dir: The SOA config directory to read from
+    :return: list of applications to kill
+    """
+    log.info("Determining apps to be killed")
+
+    applications_to_kill: List[Application] = []
+    for instance_tuple, applications in applications_dict.items():
+        service = instance_tuple[0]
+        instance = instance_tuple[1]
```

Comment:
in case you'd like to remove some indexing:
```suggestion
    for (service, instance), applications in applications_dict.items():
```
example of how this works:
```
>>> a = {(1,2): 3}
>>> for (x, y), z in a.items():
...     print(f"{x} {y} {z}")
...
1 2 3
```